sql/schemachanger: extend metadata updater to support for updating zone configs

fqazi opened the pull request:
This PR will extend the metadata updater to first support updating zone config, and next adopt
this infrastructure inside the existing zone config code.

fqazi committed:
sql/schemachanger: support setting zone configs inside MetadataUpdater

Previously, there was no way for the declarative schema changer to set
zone configs without invoking into sql, which would lead to a circular
dependency. These changes add an interface for allow the declarative
schema changes to issue zone config updates via the MetadataUpdater.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83804)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: adopt metadata updater for zone config updates

Previously, the code to update the zone configs was separate
between the declarative schema changer and the legacy schema
changer. This was inadequate because we have a opportunity
to share logic here. To address this the logic to update and
delete zone configurations is now shared.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> after you remove the unused codec

> Reviewed 20 of 20 files at r1, 17 of 17 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83804#-:-N6KSYGVBNLUNRiPeux_:b-xemsyl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @fqazi)

___
*[`pkg/sql/descmetadata/metadata_updater.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83804#-N6KS-q_0YIrNhQ7RZMw:-N6KS-q_0YIrNhQ7RZMx:b-bhf00) ([raw file](https://github.com/cockroachdb/cockroach/blob/7da053e8cefaa7dff1938e82315a99ce0bc743f8/pkg/sql/descmetadata/metadata_updater.go#L46)):*
> ```Go
> 	descriptors  *descs.Collection
> 	cacheEnabled bool
> 	codec        keys.SQLCodec
> ```

Is this used anymore?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
@ajwerner TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83804#-:-N6L2FQo-HdAV80svSuf:b-gz014k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @dt)

___
*[`pkg/sql/descmetadata/metadata_updater.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83804#-N6KS-q_0YIrNhQ7RZMw:-N6L2HWU-H2MlCihWy3W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7da053e8cefaa7dff1938e82315a99ce0bc743f8/pkg/sql/descmetadata/metadata_updater.go#L46)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is this used anymore?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
